main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala

```diff
@@ -722,18 +724,23 @@ object Stratifier extends Phase[Root, Root] {
             // Unable to stratify. Do not cache the result.
             Validation.Failure(errors)
         }
-      case Some(stf) =>
-        // Case 2: Cache hit: Return the stratification.
-        stf.toSuccess
     }
   }
 
+  private def constraintGraphToDepedencyGraph(c: ConstraintGraph): Set[DependencyEdge] =
+    c.xs.flatMap {
+      case ConstraintMultiEdge(head, positives, negatives) =>
+        val p = positives.map { case (b, loc) => DependencyEdge.Positive(head, b, loc) }
+        val n = negatives.map { case (b, loc) => DependencyEdge.Negative(head, b, loc) }
+        p ++ n
+    }
+
   /**
     * Computes the stratification of the given dependency graph `g` at the given source location `loc`.
     *
     * See Database and Knowledge - Base Systems Volume 1 Ullman, Algorithm 3.5 p 133
     */
-  private def stratify(g: DependencyGraph, tpe: Type, loc: SourceLocation): Validation[Ast.Stratification, StratificationError] = {
+  private def stratify(g: Set[DependencyEdge], tpe: Type, loc: SourceLocation): Validation[Ast.Stratification, StratificationError] = {
```

Comment:
Maybe this function should simply be moved into a separate class.

Reply:
Ullman.scala? And put the graph definition in there that ullman uses?

Reply:
UllmansAlgorithm sounds good